README
NAME
    Catalyst::Plugin::Session::State::Cookie - Maintain session IDs using
    cookies.

SYNOPSIS
        use Catalyst qw/Session Session::State::Cookie Session::Store::Foo/;

DESCRIPTION
    In order for Catalyst::Plugin::Session to work the session ID needs to
    be stored on the client, and the session data needs to be stored on the
    server.

    This plugin stores the session ID on the client using the cookie
    mechanism.

METHODS
    make_session_cookie
        Returns a hash reference with the default values for new cookies.

    update_session_cookie $hash_ref
        Sets the cookie based on "cookie_name" in the response object.

    calc_expiry
    calculate_session_cookie_expires
    cookie_is_rejecting
    delete_session_id
    extend_session_id
    get_session_cookie
    get_session_id
    set_session_id

EXTENDED METHODS
    prepare_cookies
        Will restore if an appropriate cookie is found.

    finalize_cookies
        Will set a cookie called "session" if it doesn't exist or if it's
        value is not the current session id.

    setup_session
        Will set the "cookie_name" parameter to it's default value if it
        isn't set.

CONFIGURATION
    cookie_name
        The name of the cookie to store (defaults to
        "Catalyst::Utils::apprefix($c) . '_session'").

    cookie_domain
        The name of the domain to store in the cookie (defaults to current
        host)

    cookie_expires
        Number of seconds from now you want to elapse before cookie will
        expire. Set to 0 to create a session cookie, ie one which will die
        when the user's browser is shut down.

    cookie_secure
        If this attribute set true, the cookie will only be sent via HTTPS.

    cookie_path
        The path of the request url where cookie should be baked.

CAVEATS
    Sessions have to be created before the first write to be saved. For
    example:

            sub action : Local {
                    my ( $self, $c ) = @_;
                    $c->res->write("foo");
                    $c->session( ... );
                    ...
            }

    Will cause a session ID to not be set, because by the time a session is
    actually created the headers have already been sent to the client.

SEE ALSO
    Catalyst, Catalyst::Plugin::Session.

AUTHORS
    Yuval Kogman <nothingmuch@woobling.org>

CONTRIBUTORS
    This module is derived from Catalyst::Plugin::Session::FastMmap code,
    and has been heavily modified since.

    Andrew Ford Andy Grundman Christian Hansen Marcus Ramberg Jonathan
    Rockway <jrockway@cpan.org> Sebastian Riedel

COPYRIGHT
    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Catalyst::Plugin::Session::State::Cookie

        - add SameSite support (RT#133009)
        - convert optional test using Test::WWW::Mechanize::Catalyst to
          mandatory test using Catalyst::Test
        - convert from Module::Install to Distar for packaging and release
          process
        - fix running Makefile.PL when . is not included in @INC (RT#121882)
        - minor test cleanups
        - typo and spelling fixes in documentation
        - convert repository to git (fREW Schmidt)

0.17    2009-10-18
        - Fixup copyright information

0.16    2009-10-16
        - Use session config handling from Catalyst::Plugin::Session 0.27.

0.15    2009-10-06
        - Fix the httponly option again (Closes RT##50249).
        - Make tests not warn with latest version of Catalyst.
        - Prefer session configuration to be in the 'Plugin::Session'
          config key, but provide backwards compatibility for the
          deprecated 'session' key.

0.14    2009-08-22
        - Allow turning off the httponly option (Closes RT#48930).

0.13    2009-08-19
        - Remove Test::MockObject from the test suite as prone to failing on
          some platforms and perl versions due to its UNIVERSAL:: package
          dependencies.
        - Remove Class::Accessor::Fast and replace with Moose. This allows
          us to not have a ->new method, This is more correct for Plugins
          and also means that Catalyst is not forced to invoke the scary
          replace_constructor at scope end handling.

0.12    2009-07-18
        - Introduced a new option cookie_httponly
        - Option cookie_secure extended (old syntax fully supported)

0.11    2009-05-13
        - Change TestApp so that the application is in t/lib, to make it easier
          for Catalyst to force our package to be immutable.

0.10    2009-02-08
        - POD addition.
        - Switch from NEXT to MRO::Compat

0.09    2007-10-08
        - Bump dependencies so that streaming a file also causes the cookie to
          be updated.
          - Add tests for this

0.08    2007-09-14
        - Fix live test with Mech version 0.37+
        - Switch to Module::Install

0.07    2007-04-04 23:10:00
        - Fix the bug that caused sessions to expire immediately when another
          session was deleted previously in the same request cycle

0.06    2006-09-20 19:53:00
        - Fix args in calls to NEXT::set_session_id and NEXT::extend_session_id

0.05    2006-08-06 20:50:00
        - Add configuration of 'cookie_path' (Michael W Peterson)

0.04
        - Depend on a higher version of C::P::Session

0.03
        - refactored make_session_cookie to separate calc_expiry, and made
         that easily overloadable.
        - updated for the new state API (get_session_id, set_session_id, etc)

0.02    2005-12-28 13:51:00
        - Fixed cookie_expires to support browser session cookies.
        - Renamed default cookie name from just 'session' to
          'yourapp_session' to allow several applications on the same domain
          to exist without conflict (LTJake)

0.01    2005-11-14 12:45:00
        - Initial release.


t/lib/CookieTestApp.pm
package # Hide from PAUSE
  CookieTestApp;
use Catalyst qw/
  Session
  Session::Store::Dummy
  Session::State::Cookie
  /;

__PACKAGE__->config->{session} = { cookie_secure => 2 };

sub page : Local {
    my ( $self, $c ) = @_;
    $c->res->body( "Hi! hit number " . ++$c->session->{counter} );
}

sub stream : Local {
    my ( $self, $c ) = @_;
    my $count = ++$c->session->{counter};
    $c->res->write("hit number ");
    $c->res->write($count);
}

sub deleteme : Local {
    my ( $self, $c ) = @_;
    my $id = $c->get_session_id;
    $c->delete_session;
    my $id2 = $c->get_session_id;
    $c->res->body( $id ne ( $id2 || '' ) );
}

__PACKAGE__->setup;

1;



t/lib/CookieTestApp/Controller/Root.pm
package # PAUSE HIDE
    CookieTestApp::Controller::Root;
use strict;
use warnings;

use base qw/Catalyst::Controller/;

__PACKAGE__->config( namespace => '' );

sub page : Local {
    my ( $self, $c ) = @_;
    $c->res->body( "Hi! hit number " . ++$c->session->{counter} );
}

sub stream : Local {
    my ( $self, $c ) = @_;
    my $count = ++$c->session->{counter};
    $c->res->write("hit number ");
    $c->res->write($count);
}

sub set_session_cookie_expire : Local {
    my ( $self, $c, $val ) = @_;
    $val = undef if $val eq 'undef';
    $c->set_session_cookie_expire($val);
    $c->res->body('expire_with_browser_session');
}

sub deleteme : Local {
    my ( $self, $c ) = @_;
    my $id = $c->get_session_id;
    $c->delete_session;
    my $id2 = $c->get_session_id;
    $c->res->body( $id ne ( $id2 || '' ) );
}

1;


FastMmap.pm
package Catalyst::Plugin::Session::FastMmap;

use strict;
use base qw/Class::Data::Inheritable Class::Accessor::Fast/;
use NEXT;
use Cache::FastMmap;
use Digest::MD5;
use URI;
use URI::Find;
use File::Temp 'tempdir';

our $VERSION = '0.13';

__PACKAGE__->mk_classdata('_session');
__PACKAGE__->mk_accessors('sessionid');

=head1 NAME

Catalyst::Plugin::Session::FastMmap - FastMmap sessions for Catalyst

=head1 SYNOPSIS

    use Catalyst 'Session::FastMmap';
    
    MyApp->config->{session} = {
        expires => 3600,
        rewrite => 1,
        storage => '/tmp/session'
    };

    $c->session->{foo} = 'bar';
    print $c->sessionid;

=head1 DESCRIPTION

Fast sessions.

=head2 EXTENDED METHODS

=head3 finalize

=cut

sub finalize {
    my $c = shift;
    if ( $c->config->{session}->{rewrite} ) {
        my $redirect = $c->response->redirect;
        $c->response->redirect( $c->uri($redirect) ) if $redirect;
    }
    if ( my $sid = $c->sessionid ) {
        $c->_session->set( $sid, $c->session );
        my $set = 1;
        if ( my $cookie = $c->request->cookies->{session} ) {
            $set = 0 if $cookie->value eq $sid;
        }
        if ( $set ) {
            $c->response->cookies->{session} = { 
                value => $sid
            };
        }
        if ( $c->config->{session}->{rewrite} ) {
            my $finder = URI::Find->new(
                sub {
                    my ( $uri, $orig ) = @_;
                    my $base = $c->request->base;
                    return $orig unless $orig =~ /^$base/;
                    return $orig if $uri->path =~ /\/-\//;
                    return $c->uri($orig);
                }
            );
            $finder->find( \$c->res->{body} ) if $c->res->body;
        }
    }
    return $c->NEXT::finalize(@_);
}

=head3 prepare_action

=cut

sub prepare_action {
    my $c = shift;
    if ( $c->request->path =~ /^(.*)\/\-\/(.+)$/ ) {
        $c->request->path($1);
        $c->sessionid($2);
        $c->log->debug(qq/Found sessionid "$2" in path/) if $c->debug;
    }
    if ( my $cookie = $c->request->cookies->{session} ) {
        my $sid = $cookie->value;
        $c->sessionid($sid);
        $c->log->debug(qq/Found sessionid "$sid" in cookie/) if $c->debug;
    }
    $c->NEXT::prepare_action(@_);
}

sub session {
    my $c = shift;
    return $c->{session} if $c->{session};
    my $sid = $c->sessionid;
    if (   $sid
        && $c->_session
        && ( $c->{session} = $c->_session->get($sid) ) )
    {
        $c->log->debug(qq/Found session "$sid"/) if $c->debug;
        return $c->{session};
    }
    else {
        my $sid = Digest::MD5::md5_hex( time, rand, $$, 'catalyst' );
        $c->sessionid($sid);
        $c->log->debug(qq/Created session "$sid"/) if $c->debug;
        return $c->{session} = {};
    }
}

=head3 setup

=cut

sub setup {
    my $self = shift;
    $self->config->{session}->{storage} ||= '/tmp/session';
    $self->config->{session}->{expires} ||= 60 * 60 * 24;
    $self->config->{session}->{rewrite} ||= 0;

    $self->_session(
        Cache::FastMmap->new(
            share_file  => $self->config->{session}->{storage},
            expire_time => $self->config->{session}->{expires}
        )
    );

    return $self->NEXT::setup(@_);
}

=head2 METHODS

=head3 session

=head3 uri

Extends an uri with session id if needed.

    my $uri = $c->uri('http://localhost/foo');

=cut

sub uri {
    my ( $c, $uri ) = @_;
    if ( my $sid = $c->sessionid ) {
        $uri = URI->new($uri);
        my $path = $uri->path;
        $path .= '/' unless $path =~ /\/$/;
        $uri->path( $path . "-/$sid" );
        return $uri->as_string;
    }
    return $uri;
}

=head2 CONFIG OPTIONS

=head3 rewrite

To enable automatic storing of sessions in the url set this to a true value.

=head3 storage

File to mmap for sharing of data, defaults to /tmp/session.

=head3 expires

how many seconds until the session expires. defaults to 1 day

=head1 SEE ALSO

L<Catalyst> L<Cache::FastMmap>.

=head1 AUTHOR

Sebastian Riedel, C<sri@cpan.org>
Marcus Ramberg C<mramberg@cpan.org>

=head1 COPYRIGHT

This program is free software, you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


lib/Catalyst/Plugin/Session/State/Cookie.pm
package Catalyst::Plugin::Session::State::Cookie;
use Moose;
use namespace::autoclean;

extends 'Catalyst::Plugin::Session::State';

use MRO::Compat;
use Catalyst::Utils ();

our $VERSION = '0.17';
$VERSION =~ tr/_//d;

has _deleted_session_id => ( is => 'rw' );

sub setup_session {
    my $c = shift;

    $c->maybe::next::method(@_);

    $c->_session_plugin_config->{cookie_name}
        ||= Catalyst::Utils::appprefix($c) . '_session';
}

sub extend_session_id {
    my ( $c, $sid, $expires ) = @_;

    if ( my $cookie = $c->get_session_cookie ) {
        $c->update_session_cookie( $c->make_session_cookie( $sid ) );
    }

    $c->maybe::next::method( $sid, $expires );
}

sub set_session_id {
    my ( $c, $sid ) = @_;

    $c->update_session_cookie( $c->make_session_cookie( $sid ) );

    return $c->maybe::next::method($sid);
}

sub update_session_cookie {
    my ( $c, $updated ) = @_;

    unless ( $c->cookie_is_rejecting( $updated ) ) {
        my $cookie_name = $c->_session_plugin_config->{cookie_name};
        $c->response->cookies->{$cookie_name} = $updated;
    }
}

sub cookie_is_rejecting {
    my ( $c, $cookie ) = @_;

    if ( $cookie->{path} ) {
        return 1 if index '/'.$c->request->path, $cookie->{path};
    }

    return 0;
}

sub make_session_cookie {
    my ( $c, $sid, %attrs ) = @_;

    my $cfg    = $c->_session_plugin_config;
    my $cookie = {
        value => $sid,
        ( $cfg->{cookie_domain} ? ( domain => $cfg->{cookie_domain} ) : () ),
        ( $cfg->{cookie_path} ? ( path => $cfg->{cookie_path} ) : () ),
        %attrs,
    };

    unless ( exists $cookie->{expires} ) {
        $cookie->{expires} = $c->calculate_session_cookie_expires();
    }

    #beware: we have to accept also the old syntax "cookie_secure = true"
    my $sec = $cfg->{cookie_secure} || 0; # default = 0 (not set)
    $cookie->{secure} = 1 unless ( ($sec==0) || ($sec==2) );
    $cookie->{secure} = 1 if ( ($sec==2) && $c->req->secure );

    $cookie->{httponly} = $cfg->{cookie_httponly};
    $cookie->{httponly} = 1
        unless defined $cookie->{httponly}; # default = 1 (set httponly)

    $cookie->{samesite} = $cfg->{cookie_samesite};
    $cookie->{samesite} = "Lax"
        unless defined $cookie->{ samesite}; # default = Lax

    return $cookie;
}

sub calc_expiry { # compat
    my $c = shift;
    $c->maybe::next::method( @_ ) || $c->calculate_session_cookie_expires( @_ );
}

sub calculate_session_cookie_expires {
    my $c   = shift;
    my $cfg = $c->_session_plugin_config;

    my $value = $c->maybe::next::method(@_);
    return $value if $value;

    if ( exists $cfg->{cookie_expires} ) {
        if ( $cfg->{cookie_expires} > 0 ) {
            return time() + $cfg->{cookie_expires};
        }
        else {
            return undef;
        }
    }
    else {
        return $c->session_expires;
    }
}

sub get_session_cookie {
    my $c = shift;

    my $cookie_name = $c->_session_plugin_config->{cookie_name};

    return $c->request->cookies->{$cookie_name};
}

sub get_session_id {
    my $c = shift;

    if ( !$c->_deleted_session_id and my $cookie = $c->get_session_cookie ) {
        my $sid = $cookie->value;
        $c->log->debug(qq/Found sessionid "$sid" in cookie/) if $c->debug;
        return $sid if $sid;
    }

    $c->maybe::next::method(@_);
}

sub delete_session_id {
    my ( $c, $sid ) = @_;

    $c->_deleted_session_id(1); # to prevent get_session_id from returning it

    $c->update_session_cookie( $c->make_session_cookie( $sid, expires => 0 ) );

    $c->maybe::next::method($sid);
}

1;
__END__

=head1 NAME

Catalyst::Plugin::Session::State::Cookie - Maintain session IDs using cookies.

=head1 SYNOPSIS

    use Catalyst qw/Session Session::State::Cookie Session::Store::Foo/;

=head1 DESCRIPTION

In order for L<Catalyst::Plugin::Session> to work the session ID needs to be
stored on the client, and the session data needs to be stored on the server.

This plugin stores the session ID on the client using the cookie mechanism.

=head1 METHODS

=over 4

=item make_session_cookie

Returns a hash reference with the default values for new cookies.

=item update_session_cookie $hash_ref

Sets the cookie based on C<cookie_name> in the response object.

=item calc_expiry

=item calculate_session_cookie_expires

=item cookie_is_rejecting

=item delete_session_id

=item extend_session_id

=item get_session_cookie

=item get_session_id

=item set_session_id

=back

=head1 EXTENDED METHODS

=over 4

=item prepare_cookies

Will restore if an appropriate cookie is found.

=item finalize_cookies

Will set a cookie called C<session> if it doesn't exist or if its value is not
the current session id.

=item setup_session

Will set the C<cookie_name> parameter to its default value if it isn't set.

=back

=head1 CONFIGURATION

=over 4

=item cookie_name

The name of the cookie to store (defaults to C<Catalyst::Utils::apprefix($c) . '_session'>).

=item cookie_domain

The name of the domain to store in the cookie (defaults to current host)

=item cookie_expires

Number of seconds from now you want to elapse before cookie will expire.
Set to 0 to create a session cookie, ie one which will die when the
user's browser is shut down.

=item cookie_secure

If this attribute B<set to 0> the cookie will not have the secure flag.

If this attribute B<set to 1> (or true for backward compatibility) - the cookie
sent by the server to the client will get the secure flag that tells the browser
to send this cookie back to the server only via HTTPS.

If this attribute B<set to 2> then the cookie will get the secure flag only if
the request that caused cookie generation was sent over https (this option is
not good if you are mixing https and http in your application).

Default value is 0.

=item cookie_httponly

If this attribute B<set to 0>, the cookie will not have HTTPOnly flag.

If this attribute B<set to 1>, the cookie will got HTTPOnly flag that should
prevent client side Javascript accessing the cookie value - this makes some
sort of session hijacking attacks significantly harder. Unfortunately not all
browsers support this flag (MSIE 6 SP1+, Firefox 3.0.0.6+, Opera 9.5+); if
a browser is not aware of HTTPOnly the flag will be ignored.

Default value is 1.

Note1: Many people are confused by the name "HTTPOnly" - it B<does not mean>
that this cookie works only over HTTP and not over HTTPS.

Note2: This parameter requires Catalyst::Runtime 5.80005 otherwise is skipped.

=item cookie_samesite

This attribute configures the value of the
L<SameSite|https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite>
flag.

If set to None, the cookie will be sent when making cross origin requests,
including following links from other origins. This requires the
L</cookie_secure> flag to be set.

If set to Lax, the cookie will not be included when embedded in or fetched from
other origins, but will be included when following cross origin links.

If set to Strict, the cookie will not be included for any cross origin requests,
including links from different origins.

Default value is C<Lax>. This is the default modern browsers use.

Note: This parameter requires Catalyst::Runtime 5.90125 otherwise is skipped.

=item cookie_path

The path of the request url where cookie should be baked.

=back

For example, you could stick this in MyApp.pm:

    __PACKAGE__->config( 'Plugin::Session' => {
        cookie_domain  => '.mydomain.com',
    });

=head1 CAVEATS

Sessions have to be created before the first write to be saved. For example:

    sub action : Local {
        my ( $self, $c ) = @_;
        $c->res->write("foo");
        $c->session( ... );
        ...
    }

Will cause a session ID to not be set, because by the time a session is
actually created the headers have already been sent to the client.

=head1 SEE ALSO

L<Catalyst>, L<Catalyst::Plugin::Session>.

=head1 AUTHORS

Yuval Kogman <nothingmuch@woobling.org>

=head1 CONTRIBUTORS

This module is derived from L<Catalyst::Plugin::Session::FastMmap> code, and
has been heavily modified since.

Andrew Ford

Andy Grundman

Christian Hansen

Marcus Ramberg

Jonathan Rockway <jrockway@cpan.org>

Sebastian Riedel

Florian Ragwitz

=head1 COPYRIGHT

Copyright (c) 2005 - 2009
the Catalyst::Plugin::Session::State::Cookie L</AUTHORS> and L</CONTRIBUTORS>
as listed above.

=head1 LICENSE

This program is free software, you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


Makefile.PL
use strict;
use warnings FATAL => 'all';
use 5.006;

my %META = (
  name => 'Catalyst-Plugin-Session-State-Cookie',
  license => 'perl_5',
  prereqs => {
    configure => { requires => {
      'ExtUtils::MakeMaker'   => 0,
    } },
    build => { requires => {
    } },
    test => {
      requires => {
        'Test::More'  => '0.88',
      },
    },
    runtime => {
      requires => {
        'Catalyst'                  => '5.80005',
        'Catalyst::Plugin::Session' => '0.27',
        'MRO::Compat'               => 0,
        'Moose'                     => 0,
        'namespace::autoclean'      => 0,
      },
    },
    develop   => {
      requires => {
        'Test::Pod'           => '1.14',
        'Test::Pod::Coverage' => '1.04',
      },
    },
  },
  resources => {
    repository => {
      url => 'git://git.shadowcat.co.uk/catagits/Catalyst-Plugin-Session-State-Cookie.git',
      web => 'http://git.shadowcat.co.uk/gitweb/gitweb.cgi?p=catagits/Catalyst-Plugin-Session-State-Cookie.git',
      type => 'git',
    },
    bugtracker => {
      web => 'https://rt.cpan.org/Public/Dist/Display.html?Name=Catalyst-Plugin-Session-State-Cookie',
      mailto => 'bug-Catalyst-Plugin-Session-State-Cookie@rt.cpan.org',
    },
    license => [ 'http://dev.perl.org/licenses/' ],
  },
  no_index => {
    directory => [ 't', 'xt' ]
  },
  x_authority => 'cpan:NUFFIN',
);

my %MM_ARGS = ();

## BOILERPLATE ###############################################################
require ExtUtils::MakeMaker;
(do './maint/Makefile.PL.include' or die $@) unless -f 'META.yml';

# have to do this since old EUMM dev releases miss the eval $VERSION line
my $eumm_version  = eval $ExtUtils::MakeMaker::VERSION;
my $mymeta        = $eumm_version >= 6.57_02;
my $mymeta_broken = $mymeta && $eumm_version < 6.57_07;

($MM_ARGS{NAME} = $META{name}) =~ s/-/::/g;
($MM_ARGS{VERSION_FROM} = "lib/$MM_ARGS{NAME}.pm") =~ s{::}{/}g;
$META{license} = [ $META{license} ]
  if $META{license} && !ref $META{license};
$MM_ARGS{LICENSE} = $META{license}[0]
  if $META{license} && $eumm_version >= 6.30;
$MM_ARGS{NO_MYMETA} = 1
  if $mymeta_broken;
$MM_ARGS{META_ADD} = { 'meta-spec' => { version => 2 }, %META }
  unless -f 'META.yml';
$MM_ARGS{PL_FILES} ||= {};
$MM_ARGS{NORECURS} = 1
  if not exists $MM_ARGS{NORECURS};

for (qw(configure build test runtime)) {
  my $key = $_ eq 'runtime' ? 'PREREQ_PM' : uc $_.'_REQUIRES';
  my $r = $MM_ARGS{$key} = {
    %{$META{prereqs}{$_}{requires} || {}},
    %{delete $MM_ARGS{$key} || {}},
  };
  defined $r->{$_} or delete $r->{$_} for keys %$r;
}

$MM_ARGS{MIN_PERL_VERSION} = delete $MM_ARGS{PREREQ_PM}{perl} || 0;

delete $MM_ARGS{MIN_PERL_VERSION}
  if $eumm_version < 6.47_01;
$MM_ARGS{BUILD_REQUIRES} = {%{$MM_ARGS{BUILD_REQUIRES}}, %{delete $MM_ARGS{TEST_REQUIRES}}}
  if $eumm_version < 6.63_03;
$MM_ARGS{PREREQ_PM} = {%{$MM_ARGS{PREREQ_PM}}, %{delete $MM_ARGS{BUILD_REQUIRES}}}
  if $eumm_version < 6.55_01;
delete $MM_ARGS{CONFIGURE_REQUIRES}
  if $eumm_version < 6.51_03;

ExtUtils::MakeMaker::WriteMakefile(%MM_ARGS);
## END BOILERPLATE ###########################################################
